f2py callstatement is ineffective in subroutines called from other subroutines

### Describe the issue:

When using `callstatement` to overrride a subroutine with different code as per [the docs ](https://numpy.org/doc/stable/f2py/signature-file.html#f2py-statements), this works when said subroutine is directly invoked by Python. However, when invoking another subroutine that calls the first subroutine, this override is not in effect.

MWE:
test.f90
```fortran
module utils
    implicit none
  contains
    subroutine my_abort(message)
      implicit none
      character(len=*), intent(in) :: message
      !f2py callstatement PyErr_SetString(PyExc_ValueError, message);f2py_success = 0;
      !f2py callprotoargument char*
      write(0,*) "THIS SHOULD NOT APPEAR"
      stop 1
    end subroutine my_abort

    subroutine do_something()
        call my_abort("aborting inside")
    end subroutine do_something
end module utils
```

To avoid #2547, you have to do the compilation in 2 steps:
```
python3 -m numpy.f2py -h test.pyf --no-lower -m test test.f90
python3 -m numpy.f2py -c --f90exec=gfortran --fcompiler=gnu95 test.pyf test.f90
```

Invoking `my_abort` directly works as expected, raising the ValueError I have defined:
```
$ python3 -c "import test; test.utils.my_abort('aborting directly')"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ValueError: aborting directly
```

However, when calling `do_something`, which in turn calls `my_abort`, the override seems to do nothing, and we run into the `stop`:
```
$ python3 -c "import test; test.utils.do_something()"
 THIS SHOULD NOT APPEAR
STOP 1
```
Expected result: Python raises `ValueError: aborting inside`


### Reproduce the code example:

```python
python3 -c "import test; test.utils.do_something()"
 THIS SHOULD NOT APPEAR
STOP 1

# Expected result: Python raises `ValueError: aborting inside`
```


### Error message:

_No response_

### Python and NumPy Versions:

Numpy: 2.0.0rc2 via conda-forge
python: 3.11.9 | packaged by conda-forge | (main, Apr 19 2024, 18:36:13) [GCC 12.3.0]
gfortran: GNU Fortran (conda-forge gcc 11.2.0-16) 11.2.0

Using 2.0.0rc2 to avoid issue #26156

### Runtime Environment:

[{'numpy_version': '2.0.0rc2',
  'python': '3.11.9 | packaged by conda-forge | (main, Apr 19 2024, 18:36:13) '
            '[GCC 12.3.0]',
  'uname': uname_result(system='Linux', node='b5dc6932b437', release='5.15.146.1-microsoft-standard-WSL2', version='#1 SMP Thu Jan 11 04:09:03 UTC 2024', machine='x86_64')},
 {'simd_extensions': {'baseline': ['SSE', 'SSE2', 'SSE3'],
                      'found': ['SSSE3',
                                'SSE41',
                                'POPCNT',
                                'SSE42',
                                'AVX',
                                'F16C',
                                'FMA3',
                                'AVX2',
                                'AVX512F',
                                'AVX512CD',
                                'AVX512_SKX',
                                'AVX512_CLX',
                                'AVX512_CNL',
                                'AVX512_ICL'],
                      'not_found': ['AVX512_KNL', 'AVX512_KNM', 'AVX512_SPR']}},
 {'architecture': 'SkylakeX',
  'filepath': '/opt/conda/envs/xplengine/lib/libopenblasp-r0.3.27.so',
  'internal_api': 'openblas',
  'num_threads': 4,
  'prefix': 'libopenblas',
  'threading_layer': 'pthreads',
  'user_api': 'blas',
  'version': '0.3.27'}]

### Context for the issue:

What I'm intending here is to avoid executing the Fortran `stop` statement when running wrapped code in Python, because that also kills the Python *interpreter* and thus, in my case, the `pytest` run of a whole test suite that this code is part of.

This `my_abort` cleanup subroutine is potentially called from many places throughout the actual codebase.

A similar-feeling issue involving callbacks: #7855
Feels related: #26156
ping @HaoZeke

I was thinking about this a bit. I don't really think this is something which F2PY can handle. Basically, when you call `my_abort` in `do_something()` it calls the Fortran subroutine, but the `callstatement` is only valid for calls from Python into `my_abort` (basically because the `my_abort` Python wrapper includes the `callstatement`).

This behavior is integral to the way F2PY works efficiently, since always calling Python wrapped variants of Fortran routines would be pretty inefficient.

The way to call the Python wrapped variant would be using callbacks. I'm open to more design ideas about this but as it stands this isn't actionable.

There however a bug with `callstatement` and `--lower` wherein code like this:

```
      !f2py callstatement PyErr_SetString(PyExc_ValueError, message);f2py_success = 0;
```

ends up with issues like so:

```
  640 |                 pyerr_setstring(pyexc_valueerror, message);f2py_success = 0;;
      |                 ^~~~~~~~~~~~~~~
      |                 PyErr_SetString
../cstmtmodule.c:640:33: error: 'pyexc_valueerror' undeclared (first use in this function); did you mean 'PyExc_ValueError'?
  640 |                 pyerr_setstring(pyexc_valueerror, message);f2py_success = 0;;
      |                                 ^~~~~~~~~~~~~~~~
      |                                 PyExc_ValueError
```

I'll work on that in another issue.

--- a/doc/source/f2py/python-usage.rst
+++ b/doc/source/f2py/python-usage.rst
@@ -243,6 +243,13 @@ In Python:
 .. literalinclude:: ./code/results/extcallback_session.dat
   :language: python
 
+.. note::
+
+   When using modified Fortran code via ``callstatement`` or other directives,
+   the wrapped Python function must be called as a callback, otherwise only the
+   bare Fortran routine will be used. For more details, see
+   https://github.com/numpy/numpy/issues/26681#issuecomment-2466460943
+
 Resolving arguments to call-back functions
 ------------------------------------------
 

--- a/numpy/f2py/crackfortran.py
+++ b/numpy/f2py/crackfortran.py
@@ -481,7 +481,7 @@ def readfortrancode(ffile, dowithline=show, istop=1):
                 else:
                     # clean up line beginning from possible digits.
                     l = '     ' + l[5:]
-                    # f2py directives are already by this point
+                    # f2py directives are already stripped by this point
                     if localdolowercase:
                         finalline = ll.lower()
                     else:
@@ -510,8 +510,12 @@ def readfortrancode(ffile, dowithline=show, istop=1):
                 finalline = ''
                 origfinalline = ''
             else:
-                if localdolowercase and not is_f2py_directive:
-                    finalline = ll.lower()
+                if localdolowercase:
+                    # lines with intent() should be lowered otherwise
+                    # TestString::test_char fails due to mixed case
+                    # f2py directives without intent() should be left untouched
+                    # gh-2547, gh-27697, gh-26681
+                    finalline = ll.lower() if "intent" in ll.lower() or not is_f2py_directive else ll
                 else:
                     finalline = ll
                 origfinalline = ll

--- a/numpy/f2py/crackfortran.py
+++ b/numpy/f2py/crackfortran.py
@@ -425,11 +425,14 @@ def readfortrancode(ffile, dowithline=show, istop=1):
             if l[-1] not in "\n\r\f":
                 break
             l = l[:-1]
+        # Do not lower for directives, gh-2547, gh-27697, gh-26681
+        is_f2py_directive = False
         # Unconditionally remove comments
         (l, rl) = split_by_unquoted(l, '!')
         l += ' '
         if rl[:5].lower() == '!f2py':  # f2py directive
             l, _ = split_by_unquoted(l + 4 * ' ' + rl[5:], '!')
+            is_f2py_directive = True
         if l.strip() == '':  # Skip empty line
             if sourcecodeform == 'free':
                 # In free form, a statement continues in the next line
@@ -449,6 +452,7 @@ def readfortrancode(ffile, dowithline=show, istop=1):
             if l[0] in ['*', 'c', '!', 'C', '#']:
                 if l[1:5].lower() == 'f2py':  # f2py directive
                     l = '     ' + l[5:]
+                    is_f2py_directive = True
                 else:  # Skip comment line
                     cont = False
                     continue
@@ -476,7 +480,7 @@ def readfortrancode(ffile, dowithline=show, istop=1):
                 else:
                     # clean up line beginning from possible digits.
                     l = '     ' + l[5:]
-                    if localdolowercase:
+                    if localdolowercase and not is_f2py_directive:
                         finalline = ll.lower()
                     else:
                         finalline = ll
@@ -504,7 +508,7 @@ def readfortrancode(ffile, dowithline=show, istop=1):
                 finalline = ''
                 origfinalline = ''
             else:
-                if localdolowercase:
+                if localdolowercase and not is_f2py_directive:
                     finalline = ll.lower()
                 else:
                     finalline = ll
@@ -537,7 +541,7 @@ def readfortrancode(ffile, dowithline=show, istop=1):
         else:
             dowithline(finalline)
         l1 = ll
-    if localdolowercase:
+    if localdolowercase and not is_f2py_directive:
         finalline = ll.lower()
     else:
         finalline = ll

--- a/numpy/f2py/crackfortran.py
+++ b/numpy/f2py/crackfortran.py
@@ -455,6 +455,7 @@ def readfortrancode(ffile, dowithline=show, istop=1):
                     is_f2py_directive = True
                 else:  # Skip comment line
                     cont = False
+                    is_f2py_directive = False
                     continue
             elif strictf77:
                 if len(l) > 72:
@@ -480,7 +481,8 @@ def readfortrancode(ffile, dowithline=show, istop=1):
                 else:
                     # clean up line beginning from possible digits.
                     l = '     ' + l[5:]
-                    if localdolowercase and not is_f2py_directive:
+                    # f2py directives are already by this point
+                    if localdolowercase:
                         finalline = ll.lower()
                     else:
                         finalline = ll
@@ -541,7 +543,8 @@ def readfortrancode(ffile, dowithline=show, istop=1):
         else:
             dowithline(finalline)
         l1 = ll
-    if localdolowercase and not is_f2py_directive:
+    # Last line should never have an f2py directive anyway
+    if localdolowercase:
         finalline = ll.lower()
     else:
         finalline = ll

--- a/numpy/f2py/tests/src/callback/gh26681.f90
+++ b/numpy/f2py/tests/src/callback/gh26681.f90
@@ -1,0 +1,18 @@
+module utils
+    implicit none
+  contains
+    subroutine my_abort(message)
+      implicit none
+      character(len=*), intent(in) :: message
+      !f2py callstatement PyErr_SetString(PyExc_ValueError, message);f2py_success = 0;
+      !f2py callprotoargument char*
+      write(0,*) "THIS SHOULD NOT APPEAR"
+      stop 1
+    end subroutine my_abort
+
+    subroutine do_something(message)
+        !f2py    intent(callback, hide) mypy_abort
+        character(len=*), intent(in) :: message
+        call mypy_abort(message)
+    end subroutine do_something
+end module utils

--- a/numpy/f2py/tests/src/crackfortran/gh27697.f90
+++ b/numpy/f2py/tests/src/crackfortran/gh27697.f90
@@ -1,0 +1,12 @@
+module utils
+    implicit none
+  contains
+    subroutine my_abort(message)
+      implicit none
+      character(len=*), intent(in) :: message
+      !f2py callstatement PyErr_SetString(PyExc_ValueError, message);f2py_success = 0;
+      !f2py callprotoargument char*
+      write(0,*) "THIS SHOULD NOT APPEAR"
+      stop 1
+    end subroutine my_abort
+end module utils

--- a/numpy/f2py/tests/test_callback.py
+++ b/numpy/f2py/tests/test_callback.py
@@ -244,3 +244,14 @@ class TestGH25211(util.F2PyTest):
 
         res = self.module.foo(bar)
         assert res == 110
+
+
+@pytest.mark.slow
+class TestCBFortranCallstatement(util.F2PyTest):
+    sources = [util.getpath("tests", "src", "callback", "gh26681.f90")]
+    options = ['--lower']
+
+    def test_callstatement_fortran(self):
+        with pytest.raises(ValueError, match='helpme') as exc:
+            self.module.mypy_abort = self.module.utils.my_abort
+            self.module.utils.do_something('helpme')

--- a/numpy/f2py/tests/test_callback.py
+++ b/numpy/f2py/tests/test_callback.py
@@ -5,6 +5,7 @@ import pytest
 import threading
 import traceback
 import time
+import platform
 
 import numpy as np
 from numpy.testing import IS_PYPY
@@ -247,6 +248,9 @@ class TestGH25211(util.F2PyTest):
 
 
 @pytest.mark.slow
+@pytest.mark.xfail(condition=(platform.system().lower() == 'darwin'),
+                   run=False,
+                   reason="Callback aborts cause CI failures on macOS")
 class TestCBFortranCallstatement(util.F2PyTest):
     sources = [util.getpath("tests", "src", "callback", "gh26681.f90")]
     options = ['--lower']

--- a/numpy/f2py/tests/test_crackfortran.py
+++ b/numpy/f2py/tests/test_crackfortran.py
@@ -403,3 +403,12 @@ class TestParamEval:
         dimspec = '(0:4, 3:12, 5)'
         pytest.raises(ValueError, crackfortran.param_eval, v, g_params, params,
                       dimspec=dimspec)
+
+@pytest.mark.slow
+class TestLowerF2PYDirective(util.F2PyTest):
+    sources = [util.getpath("tests", "src", "crackfortran", "gh27697.f90")]
+    options = ['--lower']
+
+    def test_no_lower_fail(self):
+        with pytest.raises(ValueError, match='aborting directly') as exc:
+            self.module.utils.my_abort('aborting directly')

